### DISCUSSION TITLE
can TxManager and RxManager switch between EUPI_MODE and DEFAULT_MODE?

### DISCUSSION THREAD

**Discussion by SeonJK** [euphony-io/euphony#241] [Q&A] [answered]
안녕하세요. EuPI TxManager 싱글톤을 생각하다가 질문이 생겼습니다!

저의 기억으로는 (정확하지 않습니다.) 챌린지 기간에 피드백을 받기로
하나의 TxManager를 EuPI모드로 전송을 수행했다가 디폴트모드로 바꿔서 전송을 수행하는게 어려움이 있을 수 있다고 하셔서 EuPI모드의 TxManager와 디폴트모드의 TxManager를 만드는 것으로 수정한 기억이 있습니다.
제가 제대로 이해하지 못하고 있을 수도 있는데 하나의 TxManager로 두가지 모드를 번갈아가며 전송하는 것에 어려움이 있다면 싱글톤을 만들면 한가지 모드만 수행할 수 있는 것이 아닌가 하는 궁금증이 생겼습니다.

그래서 저의 질문은
EuPI 모드와 디폴트 모드를 하나의 TxManager, RxManager로 번갈아가며 사용할 수 있는지 궁금합니다!

**Comment by designe** [comment, marked as answer]
안녕하세요! @SeonJK 님

EuTxManager와 EuRxManager가 내부적으로 사용하는 EuNativeConnector의 경우,
실제로는 싱글톤으로 구현되어 있어서 하나의 인스턴스만 사용하고 있습니다.

https://github.com/euphony-io/euphony/blob/ec51119dc239fa2e91846afd251cee20cf6113e5/euphony/src/main/java/co/euphony/common/EuNativeConnector.java#L30

질문하신 하나의 인스턴스를 통해 2가지 모드를 번갈아 쓸 수 있냐는 질문은 **"가능하다"** 입니다.

EuRxManager의 경우, setOption을 통해 바꿀 수 있고,
https://github.com/euphony-io/euphony/blob/ec51119dc239fa2e91846afd251cee20cf6113e5/euphony/src/main/java/co/euphony/rx/EuRxManager.java#L63

EuTxManager의 경우, setMode를 통해 바꿀 수 있습니다.
https://github.com/euphony-io/euphony/blob/ec51119dc239fa2e91846afd251cee20cf6113e5/euphony/src/main/java/co/euphony/tx/EuTxManager.java#L118

PS.
여러 인스턴스로 사용하게 되면 안정성이 떨어질 수 있기 때문에 하나의 인스턴스로 사용하는 것을 권장합니다.

예를 들어, 하나의 인스턴스에서 18500hz를 발생시켰다고 했을때
다른 인스턴스에서 18500hz가 발생하고 있는지 감지할 수가 없기 때문에 중복 발생이 될 수 있고,
EuPI와 default모드를 함께 사용하는 것도 통신에서 중복되는 frequency를 발생시킬 수 있기 때문에 
처음부터 막히는게 좋을 것 같습니다 :)

이를 위해 싱글톤이 구성되면 디자인 패턴으로 클래스의 중복 사용을 막을 수 있기 때문에 꼭 필요하다는 생각이 드네요 :)
